### ISSUE TITLE
How to make ksuid work with active storage & action text?

### ISSUE THREAD

**Issue by yourtallness** [michaelherold/ksuid-ruby#22]
Per [this guide](https://krusetech.com/2019/05/activestorage-and-uuid/), if one needs uuid support, both the active storage table and the `record_id` reference need to be of the `uuid` type.

How can I make this work with ksuid?

I'm guessing `ksuid` is not a type pg will recognize.

It also does not seem to have an impact when I append `type: :ksuid` to the `references` statement.

In the generated schema, the `record_id` is still a bigint.

Please advise

**Comment by michaelherold** [comment]
I haven't done this before so I'm guessing, but I think you should be able to follow that guide that you linked but instead of putting `id: :uuid` in the `create_table` statement, put `id: :ksuid`.

If that doesn't work, can you attach a minimal reproduction for me to look at?

**Comment by yourtallness** [comment]
Just try the following migration:

```ruby
create_table :active_storage_attachments, id: :ksuid do |t|
      t.string     :name,     null: false
      t.references :record,   null: false, polymorphic: true, index: false, type: :ksuid
      t.references :blob,     null: false

      t.datetime :created_at, null: false

      t.index [ :record_type, :record_id, :name, :blob_id ], name: "index_active_storage_attachments_uniqueness", unique: true
      t.foreign_key :active_storage_blobs, column: :blob_id
    end
```

It produces the following error:

```
PG::UndefinedObject: ERROR:  type "ksuid" does not exist
LINE 1: CREATE TABLE "active_storage_attachments" ("id" ksuid NOT NU...
```

I assume this is because `ksuid` reaches the SQL, but as far as the DB is concerned `ksuid` is just a string.

With `table.ksuid :id, primary_key: true` I can get around this but the `references` statement dues not support `type: :ksuid` either:

```
Caused by:
PG::UndefinedObject: ERROR:  type "ksuid" does not exist
LINE 1: ...ord_type" character varying NOT NULL, "record_id" ksuid NOT ...
```

**Comment by michaelherold** [comment]
Hmm. I'm running that exact migration in the test suite and it's passing, but the test suite uses in-memory SQLite. Perhaps that's the difference? I may try to make it so we test PostgreSQL and MySQL in the test suite as well. This would show in [michaelherold/activerecord-ksuid](https://github.com/michaelherold/activerecord-ksuid/), not this repository because I am splitting out the database adapter into a separate gem. See #20 for more information.

That being said, I don't think it's a good idea to do this. I think it would be better to keep the serial ID primary key for any record that you will need to attach something to with ActiveStorage/ActionText.

The reason for this is that if you make this change, _every_ model that you want to use ActiveStorage/ActionText with will require a KSUID primary key. That's likely not something that makes sense for your project.

Serial big integer IDs are cheap and exist for a reason; they are a database concern, not something that you should expose to the outside world. I recommend not reinventing the wheel here. Just because the library _can_ do something doesn't mean you _should_ do it. :smile:

**Comment by yourtallness** [comment]
@michaelherold Sorry for the delay in answering, I had to put my rails side project on ice for a while.

My intension was to completely replace serial IDs with ksuids in all models.

I didn't want a hybrid approach.

That of course means that this (or the AR compatibility) gem would need to play nicely in all cases that serial IDs / FKs are used in, which I understand is not a small effort.

**Comment by michaelherold** [comment]
I believe this was solved by #30 along with other association issues. The problem was that I misunderstood the purpose of the `ActiveRecord::Type#type` method and it was trying to send the type to the database. After #30 it should properly send either the string type or binary type, depending on the type of column you're using.

If you continue to see issues, let me know!
